### ISSUE TITLE
mismatch in event timestamp length

### ISSUE THREAD

**Issue by dougollerenshaw** [AllenInstitute/visual_behavior_analysis#700]
The length of the timestamp array in the dataset events dataframe does not match the length of the filtered_events array.

For example:
```
from visual_behavior.data_access import loading
oeid = 953443028
dataset = loading.get_ophys_dataset(oeid)
len(dataset.events.iloc[0]['timestamps'])
```

> 140014

`len(dataset.events.iloc[0]['filtered_events'])`

> 140012

Note that the length of the ophyst_timestamps attribute matches the length of the filtered_events attribute, so it would seem that the 'timestamps' attribute of the events dataframe is the outlier.

`len(dataset.ophys_timestamps)`
> 140012

**Comment by dougollerenshaw** [comment]
I just checked another session and it's also off by two:

```
from visual_behavior.data_access import loading
oeid = 958435448
dataset = loading.get_ophys_dataset(oeid)
len(dataset.events.iloc[0]['timestamps'])
```

> 140058

`len(dataset.events.iloc[0]['filtered_events'])`

> 140056

`len(dataset.ophys_timestamps)`

> 140056

**Comment by dougollerenshaw** [comment]
One more clue: the timestamps align at the beginning of the array, but not at the end. That'd seem to imply that the extra two timestamps in dataset.events.iloc[0]['timestamps'] are at the end:
![image](https://user-images.githubusercontent.com/19944442/102936092-59405380-445c-11eb-9392-9b9a4dc69f99.png)

**Comment by dougollerenshaw** [comment]
@matchings are those timestamps inherited directly from the array that @ledochowitsch is saving to disk?

**Comment by matchings** [comment]
@dougollerenshaw Yes. For now, don't use the timestamps in the events df. Use dataset.ophys_timestamps. Those are the ground truth from SDK. I am not sure what could be causing the timestamps in the event detection output to be off.

**Comment by dougollerenshaw** [comment]
More evidence that the extra two timestamps at the end are extraneous:
```
from visual_behavior.data_access import loading
import matplotlib.pyplot as plt

oeid = 953443028
dataset = loading.get_ophys_dataset(oeid)

fig,ax=plt.subplots()
ax.plot(
    dataset.ophys_timestamps,
    dataset.events.iloc[4]['filtered_events']
)
```
Plotting with last two timestamps trimmed off:
```
ax.plot(
    dataset.events.iloc[4]['timestamps'][:-2], #trim off last two
    dataset.events.iloc[4]['filtered_events'],
    linestyle = ':',
    linewidth = 3
)

ax.set_xlim(352,353)
ax.set_ylim(-0.01,0.05)
```
gives us two aligned traces:

![image](https://user-images.githubusercontent.com/19944442/102939597-614fc180-4463-11eb-918a-f9d4a8ab4e63.png)

But trimming off the first two gives us misaligned traces:
```
fig,ax=plt.subplots()
ax.plot(
    dataset.ophys_timestamps,
    dataset.events.iloc[4]['filtered_events']
)

ax.plot(
    dataset.events.iloc[4]['timestamps'][2:], # trim off first two
    dataset.events.iloc[4]['filtered_events'],
    linestyle = ':',
    linewidth = 3
)

ax.set_xlim(352,353)
ax.set_ylim(-0.01,0.05)
```
![image](https://user-images.githubusercontent.com/19944442/102939670-880df800-4463-11eb-8241-0de6fb64bdec.png)

**Comment by ledochowitsch** [comment]
Hey guys,

Which time stamps are you guys talking about?

Note that once you load the npz file, there are different sets of time_stamps in there:

there is npz[‘ts’], which should be identical to what you get from the SDK because it’s just the result of

dataset = loading.get_ophys_dataset(eval(exp_id), include_invalid_rois=False)
ts = dataset.timestamps.ophys_frames.values[0]

Then inside the event_dict object there is the key ‘ts’, which contains the time stamps for the detected events, for each cell id: npz[‘event_dict’][cid][‘ts’]

Those time stamps are computed by upsampling the original time stamps:

ts30Hz = resample_poly(ts, uf, 1)

figuring out where there are events in the upsampled events trace:

event_idx = np.where(event30Hz>0)[0] #upsampled

and finally indexing with that into the upsampled time stamp trace:

event_ts = ts30Hz[event_idx]

What appears to be wrong?

Best,

-Peter

**Comment by matchings** [comment]
i believe the timestamps Doug is talking about are the npz['ts'] ones, not the ones in the event_dict.

**Comment by ledochowitsch** [comment]
That’s very mysterious then – I’m just passing those through…

-Peter

**Comment by dougollerenshaw** [comment]
Thanks @ledochowitsch. Sorry for being unclear about the underlying issue. I was initially struggling to understand it myself so this issue got a little muddied. 

But the fundamental issue is this: the timestamps associated with the events are two values longer than the events arrays themselves.

For the same oeid I initially referenced in this issue, here's what happens when I go back to the cached events file:

```
import numpy as np
events_file = '//allen/programs/braintv/workgroups/nc-ophys/visual_behavior/event_detection/953443028.npz'
f = np.load(events_file, allow_pickle=True)
```
Get the length of the timestamps array:
`len(f['ts'])`

> 140014

Get the length of an events trace:
```
event_dict = f['event_dict'].item()
cell_roi_ids = list(event_dict.keys())
len(event_dict[cell_roi_ids[0]]['event_trace'])
```

> 140012

Above you said:

> there is npz[‘ts’], which should be identical to what you get from the SDK

When I go back to the directly to the SDK, I get this:
```
from allensdk.brain_observatory.behavior.behavior_ophys_session import BehaviorOphysSession
oeid = 953443028
session = BehaviorOphysSession.from_lims(oeid)
len(session.ophys_timestamps)
```
> 140012

But you went on to say:

> ...because it’s just the result of
> 
> dataset = loading.get_ophys_dataset(eval(exp_id), include_invalid_rois=False)
> ts = dataset.timestamps.ophys_frames.values[0]

Checking that myself, I see:
```
from visual_behavior.data_access import loading
dataset = loading.get_ophys_dataset(oeid, include_invalid_rois=False)
len(dataset.timestamps.ophys_frames['timestamps'])
```
> 140014

So it'd seem that the `dataset.timestamps.ophys_frames['timestamps']` attribute is the source of the confusion here. @matchings, do you know where that attribute is coming from and why it would be two elements longer than `ophys_timestamps`?

**Comment by ledochowitsch** [comment]
Ah, I see... That’s frustrating:/.

The good news is that the contents of npz[‘events’] will be unaffected by this issue. However, the events time stamps will be off by the same two samples…

-Peter

**Comment by matchings** [comment]
dataset.timestamps.ophys_frames['timestamps'] are computed directly from the sync file, and these are what is used to create dataset.ophys_timestamps for mesoscope experiments because the SDK does not yet do the proper time resampling for mesoscope (or at least it didnt in the version we are using). For scientifica, dataset.ophys_timestamps is pulled directly from the SDK. If the SDK is doing some truncation of frames, it could lead to a discrepancy between dataset.ophys_timestamps and dataset.timestamps.ophys_frames['timestamps']. But that should be specific to Scientifica, because mesoscope uses the same thing for both. I hope that makes sense...

**Comment by dougollerenshaw** [comment]
Thanks @matchings. It looks like you're correct that this is specific to scientific sessions. Here's an example from mesoscope showing that both the SDK ophys_timestamps and the VBA dataset  ophys_frames['timestamps'] vectors are the same length:

![image](https://user-images.githubusercontent.com/19944442/102944640-37e96280-4470-11eb-86b8-31cdc4f62fa3.png)

And here's a different 2P3 (scientifica) session with the same off-by-two error as above:

![image](https://user-images.githubusercontent.com/19944442/102944740-81d24880-4470-11eb-90db-cacff02155c1.png)

So does this mean that the problem is with the SDK? If so, we should submit an SDK issue to solve it. These discrepancies will undoubtedly confuse other users in the future.

**Comment by matchings** [comment]
im guessing that the SDK truncates the timestamps to match the ophys traces, which is probably a desired behavior, otherwise we would have mismatches all over the place. i believe the scientificas are known to give out a few extra TTL pulses at the end of the session (or at least MPE says its at the end, its nice that you just validated that here), which we want to remove so that everything is aligned. it surprises me that you are always seeing an off by exactly 2 issue though, because i thought the number of those extra pulses at the end was variable.

**Comment by alexpiet** [comment]
I'm always paranoid about convolutions. Should the mode be "valid" instead of "full" (default)?

https://github.com/AllenInstitute/visual_behavior_analysis/blob/0b07d4657b80431b328122efc6ef60122306b654/visual_behavior/ophys/response_analysis/response_processing.py#L422

https://numpy.org/doc/stable/reference/generated/numpy.convolve.html

**Comment by ledochowitsch** [comment]
I had in fact double-checked that the tine stamps were the same length as the dff traces when I prototyped the code - on MesoScope data. Unfortunately, I have not re-checked when I generalized it to also work for Scientifica. Who would have thunk?

-Peter

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
